packages/core-data/src/fetch/fetch-link-suggestions.ts

```diff
@@ -0,0 +1,356 @@
+							title:
+								decodeEntities( result.title || '' ) ||
+								__( '(no title)' ),
+							type: result.subtype || result.type,
+							kind: 'taxonomy',
+						};
+					} );
+				} )
+				.catch( () => [] ) // Fail by returning no results.
+		);
+	}
+
+	if ( ! type || type === 'attachment' ) {
+		queries.push(
+			apiFetch< MediaAPIResult[] >( {
+				path: addQueryArgs( '/wp/v2/media', {
+					search,
+					page,
+					per_page: perPage,
+				} ),
+			} )
+				.then( ( results ) => {
+					return results.map( ( result ) => {
+						return {
+							id: result.id,
+							url: result.source_url,
+							title:
+								decodeEntities( result.title.rendered || '' ) ||
+								__( '(no title)' ),
+							type: result.type,
+							kind: 'media',
+						};
+					} );
+				} )
+				.catch( () => [] ) // Fail by returning no results.
+		);
+	}
+
+	const responses = await Promise.all( queries );
+
+	let results = responses.flat();
+	results = results.filter( ( result ) => !! result.id );
+	results = sortResults( results, search );
+	results = results.slice( 0, perPage );
+	return results;
+}
+
+/**
+ * Sort search results by relevance to the given query.
+ *
+ * Sorting is necessary as we're querying multiple endpoints and merging the results. For example
+ * a taxonomy title might be more relevant than a post title, but by default taxonomy results will
+ * be ordered after all the (potentially irrelevant) post results.
+ *
+ * We use cosine similarity to determine relevance. This is a common technique used in information
+ * retrieval to determine how similar two documents are. In this case, we're treating the search
+ * query as a document and the title of the search result as another document.
+ *
+ * @param results
+ * @param search
+ */
+export function sortResults( results: SearchResult[], search: string ) {
+	const searchTerms = tokenize( search );
+	const searchTermFrequencies = getTermFrequencies( searchTerms );
+
+	const scores = {};
+	for ( const result of results ) {
+		const titleTerms = tokenize( result.title );
+		const titleTermFrequencies = getTermFrequencies( titleTerms );
+		scores[ result.id ] = getCosineSimilarity(
+			searchTermFrequencies,
+			titleTermFrequencies
+		);
+	}
+
+	return results.sort( ( a, b ) => scores[ b.id ] - scores[ a.id ] );
+}
+
+/**
+ * Turns text into an array of tokens.
+ *
+ * For example, `"I'm having a ball"` becomes `[ "im", "having", "a", "ball" ]`.
+ *
+ * @param text
+ */
+function tokenize( text: string ) {
+	return text
+		.toLowerCase()
+		.split( /\s+/ )
+		.map( ( word ) => word.replace( /\W/g, '' ) )
+		.filter( ( word ) => !! word );
+}
+
+/**
+ * Returns a map of term frequencies.
+ *
+ * For example "the cat in the hat" becomes `{ the: 2, cat: 1, in: 1, hat: 1 }`.
+ *
+ * The map can be interpreted as a vector in a multi-dimensional space, where
+ * each dimension corresponds to a unique word, and the value of the dimension
+ * is the frequency of that word.
```

Comment:
Yeah so it **is** a dictionary that maps words to their respective quantities. This dictionary can _also_ be viewed as a multi-dimensional vector where each dimension is a unique word and the value of that dimension is the quantity.

So `'brown brown fox'` becomes `{ brown: 2, fox: 1 }` which can be thought of as the 2d vector of `(2, 1)`.

Pretending that the dictionary is a vector is the magic trick that lets us compare how similar two dictionaries are by calculating the angle between the two vectors. (but we actually do cosine of the angle because that's easier to calculate using vector operations: dot product, magnitude.)

I'll try to update the comments to explain better 😀

Reply:
Ah, thanks! That helps clarify it for me 👍

Reply:
Trying a simpler scoring function so maybe don't need this.